### PR TITLE
Fixing issues with the elemental sampler

### PR DESCRIPTION
**Pull request by bannanc**
This should have been caught in pull request #190 

I'm looking at the tests being run on the elemental sampler to see if I can figure out why this wasn't caught earlier.

### CODE REVIEW

**Comment by davidlmobley** [comment]
@bannanc - what's the status on this?

**Comment by bannanc** [comment]
@davidlmobley 
Well the typo that lead to this is fixed, but there seems to be some other possible issues with the elemental sampler. So I think until I can completely debug it we should hold off, I'll write out a detailed issue later today. Assuming no one else is trying to use it right now.

**Comment by bannanc** [comment]
I've actually discovered some significant problems with how the elemental sampler is scoring. Details are in issue #202

**Comment by bannanc** [comment]
@davidlmobley 
when you have time could you do a code review here? 

Please note there is now only one smarty sampler, I added an optional input element. I had a rework a couple of the methods, but now it is definitely consistent sampling rather we look at all atomtypes or 1 specific element. 

I'm going to update the README in the data/molecules/ directory and make sure there are clear examples with both decorator options and how to indicate element examples.

**Comment by davidlmobley** [comment]
@bannanc - thanks! This sounds great. I'll review most likely in the morning.

File: smarty/sampler.py
Comment (davidlmobley) [inline smarty/sampler.py]: "iteger" -> "integer".

File: smarty/sampler.py
Comment (davidlmobley) [inline smarty/sampler.py:194]: Ah, clever. Well done.

File: smarty/sampler.py
Comment (davidlmobley) [inline smarty/sampler.py:285]: This is very nice.

File: smarty/sampler.py
Comment (davidlmobley) [inline smarty/sampler.py:401]: "Replacements"? Like macros? Or the recursive smarts?
Does this mean there is a "to do" item here which warrants an issue?

File: smarty/sampler.py
Comment (davidlmobley) [inline smarty/sampler.py]: I agree with you here, @bannanc .

File: smarty/tests/test_sampler.py
Comment (davidlmobley) [inline smarty/tests/test_sampler.py]: Wait, isn't "element" a keyword argument in this function, e.g. the argument should be `element=1` not just `1`??

**Review by davidlmobley — APPROVED** [review]
Basically looks good, @bannanc . Once you have your README.md files and examples cleaned up as you mentioned, and you review my (mostly minor) comments and make any changes you deem necessary this is good to merge.

File: smarty/sampler.py
Comment (bannanc) [reply smarty/sampler.py:194]: Thanks

File: smarty/sampler.py
Comment (bannanc) [reply smarty/sampler.py:401]: @davidlmobley 
I was going to check this before writing an issue. Since we aren't currently using any of the replacement strings in our initial types I didn't think it needed to be addressed in this pull request.

File: smarty/sampler.py
Comment (bannanc) [reply smarty/sampler.py]: @davidlmobley so should I just remove the method form the script all together?

File: smarty/tests/test_sampler.py
Comment (bannanc) [reply smarty/tests/test_sampler.py]: @davidlmobley 
The function is element=0 as a default, I thought if you called a function with the keyword arguments in order you didn't have put the name in front? 
Is there a style preference for this?
The decorator type, temperature, verbose setting, and replacements are also keyword arguments.

File: smarty/sampler.py
Comment (davidlmobley) [reply smarty/sampler.py:401]: Sounds good.

File: smarty/sampler.py
Comment (davidlmobley) [reply smarty/sampler.py]: Yes, IMO.

File: smarty/tests/test_sampler.py
Comment (davidlmobley) [reply smarty/tests/test_sampler.py]: Er, I thought the test here said you were trying to test elemental sampling. My mistake, if not. But, you SHOULD test elemental sampling somewhere, right?

Comment (bannanc) [reply smarty/tests/test_sampler.py]: @davidlmobley this is testing elemental sampling, the element is the last keyword argument so the 1 on the end sets element=1 (instead of 0). 

I'm confused about why I would have to specify element= when calling the function if I'm specifying all of the keyword arguments.

Comment (davidlmobley) [reply smarty/tests/test_sampler.py]: Ah, I see. OK, stylistically, I find it easier to read what's going on if you say `element=` for an optional argument... But, if that complicates too much you can leave it this way. Your call.

Comment (bannanc) [reply smarty/tests/test_sampler.py]: OK, that makes more sense, I can change it so its consistent.

Comment (davidlmobley) [reply smarty/tests/test_sampler.py]: For me it's mostly an issue about argument count. As the argument list gets long and some of them are optimal, I find it hard to track by eyeball which is which unless you use the keywords.

**Comment by bannanc** [comment]
I've made the changes David suggested, assuming I didn't break the test, I'll merge after these tests pass.

**Comment by bannanc** [comment]
Never mind, I took a closer look at my test results from last night, it looks like its still only saving basetypes that are for the element being sampled, it is supposed to save all that match any atom.

**Comment by bannanc** [comment]
There is a molecule in MiniDrugBank that can't be typed by the base-elements, I'm still not quite sure why, I'm going to investigate in the morning. 
If it is an issue with the molecule, I'll update the MiniDrugBank set (AGAIN), but maybe I'm missing something in the basetypes-elemental.smarts file.

**Comment by davidlmobley** [comment]
Interesting, @bannanc ! Keep me posted.

**Comment by bannanc** [comment]
This most recent test failed because the H3 on the "answers" simulation was removed in the only step. I changed the answers test temperature to 0 to prevent this happening in the future. 

I'm looking at MiniDrugBank and it looks like all of the molecules are typed with the elemental basetypes. I'm going to double check the input files, but I thought they were the same as the one I'm testing with.

**Comment by bannanc** [comment]
So, when I run the atomtyper in interactive mode, it can type all molecules in the MiniDrugBank_atyped.oeb file, here is what I called (mirroring how smarty types molecules):
```
from smarty import *
molecules = utils.read_molecules('MiniDrugBank_atyped.oeb')
basetypes = AtomTyper.read_typelist('atomtypes/basetypes-elemental.smarts')
atomtyper = AtomTyper(basetypes, 'atomtype')

# This works without an error
for mol in molecules:
    atomtyper.assignTypes(mol, 0)

# This prints nothing
for mol in molecules:
    for atom in mol.GetAtoms():
        if atom.GetStringData('atomtype') == "":
            print("Found untyped atom")
```

But when I run the smarty calculator with the same in put files I get a `smarty.atomtyper.TypingException`

I'm still investigating why smarty wouldn't work when the interactive does.

**Comment by bannanc** [comment]
@davidlmobley here's an update...

**Comment by davidlmobley** [comment]
@bannanc - are you doing anything different on read, or is there some cleanup being done on the molecules first? Or, interactively, are you somehow loading a different version of smarty? (Perhaps one is being done on the cluster?)

**Comment by bannanc** [comment]
I don't think so, I was interactive on the cluster so I was running the same smarty version. I'm going to remove smarty and re-install in on green planet and run again.

**Comment by bannanc** [comment]
@davidlmobley I found the problem, 

I thought this error was happening with smarty when assigning basetypes, it actually happens earlier than that.

In this round I added a section, to try to limit the bond types to those that are actually used in the molecule set. It appears there might be a chlorine in this molecule that isn't bound to anything. There is apparently a loose Chlorine ion in the molecule, here's a pdf image of the molecule.

[MiniDrugBank_113.pdf](https://github.com/open-forcefield-group/smarty/files/781703/MiniDrugBank_113.pdf)

**Comment by davidlmobley** [comment]
Yes, that would do it. Nice work.

**Comment by bannanc** [comment]
@davidlmobley This is all working now, merging. 

It was tedious and more time consuming than I expected, but I think our scripts and sets of molecules are better for it.

**Comment by davidlmobley** [comment]
Thanks, @bannanc . I also removed the [WIP] from the title. :)

**Pull request merged**
